File: sdk/java/src/test/java/cloud/oresoftware/nextloggers/NextLoggersTest.java
package cloud.oresoftware.nextloggers;

import java.util.ArrayList;
import java.util.List;
import java.util.Map;
import java.util.concurrent.Callable;
import java.util.concurrent.ExecutorService;
import java.util.concurrent.Executors;
import java.util.concurrent.Future;
import java.util.concurrent.TimeUnit;

public final class NextLoggersTest {
  private NextLoggersTest() {}

  public static void main(String[] args) throws Exception {
    List<Map<String, Object>> otel = new ArrayList<>();
    List<Map<String, Object>> supabase = new ArrayList<>();
    NextLoggers.Logger logger =
        new NextLoggers.Logger(
            "payments",
            "audit",
            "java",
            Map.of("environment", "test"),
            List.of(
                new NextLoggers.OtelTransport(otel::add),
                new NextLoggers.SupabaseTransport(supabase::add)));

    Map<String, Object> record =
        NextLoggers.withContext(
            new NextLoggers.Context(
                "0123456789abcdef0123456789abcdef",
                "0123456789abcdef",
                1,
                "vendor=value",
                Map.of("requestId", "request-1"),
                List.of("otel", "checkout")),
            () -> logger.error("payment failed", Map.of("orderId", "order-42")));

    assert NextLoggers.SCHEMA.equals(record.get("schema"));
    assert "payments".equals(record.get("appName"));
    assert "ERROR".equals(record.get("level"));
    assert "0123456789abcdef0123456789abcdef".equals(record.get("traceId"));
    @SuppressWarnings("unchecked")
    Map<String, Object> fields = (Map<String, Object>) record.get("fields");
    assert "0123456789abcdef".equals(fields.get("otel.span_id"));
    assert "request-1".equals(fields.get("requestId"));
    assert "order-42".equals(fields.get("orderId"));
    assert otel.size() == 1;
    assert Integer.valueOf(17).equals(otel.get(0).get("severityNumber"));
    assert supabase.size() == 1;
    assert NextLoggers.toJson(record).contains("\"schema\":\"next-loggers/v1\"");
    assert NextLoggers.currentContext() == null : "context was not restored";

    ExecutorService pool = Executors.newFixedThreadPool(2);
    try {
      List<Callable<String>> calls =
          List.of(
              () ->
                  NextLoggers.withContext(
                      NextLoggers.Context.trace("trace-a", "span-a"),
                      () -> String.valueOf(logger.info("a", Map.of()).get("traceId"))),
              () ->
                  NextLoggers.withContext(
                      NextLoggers.Context.trace("trace-b", "span-b"),
                      () -> String.valueOf(logger.info("b", Map.of()).get("traceId"))));
      List<Future<String>> futures = pool.invokeAll(calls);
      assert "trace-a".equals(futures.get(0).get());
      assert "trace-b".equals(futures.get(1).get());
    } finally {
      pool.shutdownNow();
      if (!pool.awaitTermination(5, TimeUnit.SECONDS)) {
        throw new AssertionError("executor did not terminate");
      }
    }

    List<Map<String, Object>> routedOtel = new ArrayList<>();
    List<Map<String, Object>> routedAll = new ArrayList<>();
    NextLoggers.Logger routed =
        new NextLoggers.Logger(
            "checkout",
            List.of(
                new NextLoggers.OtelTransport(routedOtel::add),
                new NextLoggers.SupabaseTransport(routedAll::add)));
    routed.info("default on", Map.of());
    routed.notOtel().warn("opted out", Map.of());
    routed.notOtel().useOtel().error("opted back in", Map.of());
    assert routedOtel.size() == 2 : "notOtel() must skip the OTEL transport";
    assert "default on".equals(routedOtel.get(0).get("body"));
    assert "opted back in".equals(routedOtel.get(1).get("body"));
    assert routedAll.size() == 3 : "other transports must receive every record";
    assert routed.otelEnabled() : "notOtel() must return a derived logger";

    System.out.println("Java next-loggers conformance passed");
  }
}

File: sdk/ruby/test/next_loggers_test.rb
# frozen_string_literal: true

require "minitest/autorun"
require_relative "../lib/oresoftware/next_loggers"

class NextLoggersTest < Minitest::Test
  def test_context_flows_to_otel_and_supabase_and_is_restored
    otel = []
    supabase = []
    logger = ORESoftware::NextLoggers::Logger.new(
      app_name: "payments",
      name: "audit",
      fields: { environment: "test" },
      id_factory: -> { "ruby-record-1" },
      clock: -> { "2026-01-02T03:04:05.000Z" },
      transports: [
        ORESoftware::NextLoggers::OtelTransport.new { |record| otel << record },
        ORESoftware::NextLoggers::SupabaseTransport.new { |record| supabase << record }
      ]
    )

    record = ORESoftware::NextLoggers.with_context(
      trace_id: "0123456789abcdef0123456789abcdef",
      span_id: "0123456789abcdef",
      trace_flags: 1,
      trace_state: "vendor=value",
      fields: { requestId: "request-1" },
      tags: %w[otel ruby]
    ) do
      logger.error("payment failed", orderId: "order-42")
    end

    assert_equal "next-loggers/v1", record.fetch("schema")
    assert_equal "ERROR", record.fetch("level")
    assert_equal "0123456789abcdef0123456789abcdef", record.fetch("traceId")
    assert_equal "0123456789abcdef", record.fetch("fields").fetch("otel.span_id")
    assert_equal "request-1", record.fetch("fields").fetch("requestId")
    assert_equal "order-42", record.fetch("fields").fetch("orderId")
    assert_equal 17, otel.fetch(0).fetch("severityNumber")
    assert_same record, supabase.fetch(0)
    assert_nil ORESoftware::NextLoggers.current_context
  end

  def test_thread_local_context_is_isolated
    logger = ORESoftware::NextLoggers::Logger.new(app_name: "app", transports: [])
    traces = Queue.new

    threads = { a: "trace-a", b: "trace-b" }.map do |name, trace_id|
      Thread.new do
        value = ORESoftware::NextLoggers.with_context(trace_id: trace_id, span_id: "#{name}-span") do
          logger.info(name.to_s).fetch("traceId")
        end
        traces << value
      end
    end
    threads.each(&:join)

    assert_equal %w[trace-a trace-b], [traces.pop, traces.pop].sort
  end

  def test_per_event_otel_routing_preserves_regular_transports
    otel = []
    regular = []
    logger = ORESoftware::NextLoggers::Logger.new(
      app_name: "routing",
      otel: false,
      transports: [
        ORESoftware::NextLoggers::OtelTransport.new { |record| otel << record },
        ->(record) { regular << record }
      ]
    )

    default_off = logger.event(:info, "default-off")
    refute default_off.otel_enabled?(logger.otel_enabled?)
    default_off.send
    logger.event(:info, "forced-on").use_otel.send
    logger.event(:info, "reset-off").use_otel.reset_otel.send
    logger.use_otel
    logger.event(:warn, "forced-off").not_otel.send
    logger.event(:info, "logger-on").with_otel(true).send

    assert_equal %w[forced-on logger-on], otel.map { |record| record.fetch("body") }
    assert_equal(
      %w[default-off forced-on reset-off forced-off logger-on],
      regular.map { |record| record.fetch("message") }
    )
  end
end
